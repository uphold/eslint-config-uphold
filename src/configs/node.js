/**
 * Module dependencies.
 *
 * @typedef {import('eslint').Linter.Config} LinterConfig
 */

import { defineConfig } from 'eslint/config';
import babelParser from '@babel/eslint-parser';
import globals from 'globals';
import upholdBaseConfig from './base.js';

/**
 * Configuration for bin and scripts files.
 *
 * @type {LinterConfig}
 */

const upholdBinScriptsConfig = {
  files: ['**/bin/**', '**/scripts/**'],
  name: 'uphold/scripts',
  rules: {
    'no-console': 'off',
    'node-plugin/no-process-exit': 'off'
  }
};

/**
 * Shared configuration preset for `node`.
 */

const upholdNode = defineConfig([
  {
    extends: [upholdBaseConfig, upholdBinScriptsConfig],
    languageOptions: {
      ecmaVersion: 2020,
      globals: {
        ...globals.jasmine,
        ...globals.jest,
        ...globals.mocha,
        ...globals.node
      },
      parser: babelParser,
      parserOptions: {
        requireConfigFile: false
      },
      sourceType: 'module'
    },
    name: 'uphold/node'
  }
]);

/**
 * Export the configuration.
 *
 * @see https://nodejs.org/docs/latest-v22.x/api/modules.html#loading-ecmascript-modules-using-require
 */

export { upholdNode as 'module.exports' };
export default upholdNode;
