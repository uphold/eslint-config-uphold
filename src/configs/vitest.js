/**
 * Module dependencies.
 */

import { isModuleAvailable, loadModule } from '../utils/load-module.js';

/**
 * Check if `@vitest/eslint-plugin` and its dependencies are available.
 */

const isTypeScriptAvailable = isModuleAvailable('typescript');
const isVitestPluginAvailable = isModuleAvailable('@vitest/eslint-plugin');
const isVitestAvailable = isModuleAvailable('vitest');

/**
 * Uphold Vitest ESLint config.
 * @type {import('@eslint/config-helpers').ConfigWithExtends}
 */
let upholdVitestConfig;

if (isVitestPluginAvailable && isTypeScriptAvailable && isVitestAvailable) {
  const { default: vitestPlugin } = await loadModule('@vitest/eslint-plugin');

  upholdVitestConfig = {
    extends: [vitestPlugin.configs.recommended],
    languageOptions: vitestPlugin.configs.env.languageOptions,
    name: 'uphold/vitest-plugin-config',
    rules: {
      'vitest/no-commented-out-tests': 'warn',
      'vitest/no-disabled-tests': 'warn',
      'vitest/no-focused-tests': 'error',
      'vitest/no-interpolation-in-snapshots': 'error',
      'vitest/no-mocks-import': 'error',
      'vitest/no-standalone-expect': 'error',
      'vitest/no-test-prefixes': 'error',
      'vitest/prefer-expect-resolves': 'warn',
      'vitest/prefer-to-have-length': 'warn'
    }
  };
} else if (isVitestAvailable) {
  if (!isTypeScriptAvailable) {
    console.warn('`typescript` is not installed, Vitest linting will be disabled');
  }

  if (!isVitestPluginAvailable) {
    console.warn('`@vitest/eslint-plugin` is not installed, Vitest linting will be disabled');
  }

  const globals = await loadModule('globals');

  upholdVitestConfig = {
    languageOptions: {
      globals: globals.vitest
    },
    name: 'uphold/vitest-globals'
  };
} else {
  console.error('Vitest is not installed, this config will be empty');

  upholdVitestConfig = {
    name: 'uphold/vitest-empty'
  };
}

/**
 * Export the configuration.
 */

export { upholdVitestConfig as 'module.exports' };
export default upholdVitestConfig;
