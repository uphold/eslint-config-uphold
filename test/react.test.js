/**
 * Module dependencies.
 */

import { ESLint } from 'eslint';
import { describe, expect, it } from 'vitest';
import { resolve } from 'node:path';
import baseConfig from '../src/react.js';

/**
 * Instances.
 */

const linter = new ESLint({ baseConfig, ignore: false });

/**
 * Tests `eslint-config-uphold/react`.
 */

describe('eslint-config-uphold/react', () => {
  it('should not generate any violations for `react/prefer-stateless-function` in correct code', async () => {
    const source = resolve(import.meta.dirname, 'fixtures/react/correct.js');
    const [result] = await linter.lintFiles([source]);

    expect(result.errorCount).toEqual(0);
  });

  it('should generate violations for `react/prefer-stateless-function` in incorrect code', async () => {
    const source = resolve(import.meta.dirname, 'fixtures/react/incorrect.js');
    const [result] = await linter.lintFiles([source]);

    const violations = result.messages.map(({ column, line, ruleId, severity }) => ({
      column,
      line,
      ruleId,
      severity
    }));

    expect(violations).toMatchSnapshot();
  });
});
