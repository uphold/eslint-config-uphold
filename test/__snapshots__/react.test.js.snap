// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`eslint-config-uphold/react > should generate violations for \`react/prefer-stateless-function\` in incorrect code 1`] = `
[
  {
    "column": 5,
    "line": 13,
    "ruleId": "react-hooks/rules-of-hooks",
    "severity": 2,
  },
  {
    "column": 18,
    "line": 23,
    "ruleId": "arrow-body-style",
    "severity": 2,
  },
  {
    "column": 4,
    "line": 25,
    "ruleId": "prettier/prettier",
    "severity": 2,
  },
  {
    "column": 29,
    "line": 35,
    "ruleId": "prettier/prettier",
    "severity": 2,
  },
]
`;
